### ISSUE TITLE
Exportable GraphinProvider & useGraphin React hook to access from GraphinContext

### ISSUE THREAD

**Issue by cliffordfajardo** [antvis/Graphin#333]
## Problem Statement
Currently, there is no way to access the `GraphinContext` data outside of the `<Graphin>` component.
This is because we don't expose as an export-able `GraphinProvider` component.

#### Example / Scenario 1
Imagine I have a component called `<NetworkTopologyViewer>` that would like to return a `<Graphin>` component like this:

```tsx
/**
* @description
* Renders a Graph wrapped in a custom container
* and also sets up custom node behavior
*/
const NetworkTopologyViewer = () => {
    const { graph } = React.useContext(GraphinContext); //❗ does not work. GraphinContext is not inside a Provider
    const handleNodeClick = e => {
          console.log(e.item.getModel(); 
         makeAPICall(e.item.getModel()))
    }
   
    // ❗ Cannot setup custom node behavior here because `graph` cannot be grabbed from context above
    graph.on('node:click', handleNodeClick)
  
	return (
	 <Card title="Network Topology">
		   <Graphin {...GraphinProps....}/>
      <Card>
  )
}
```

<br />
In this new proposed scenario below , we would be able to do this:

```tsx
// --------------------App.tsx--------
// Initialize with no data
import { GraphinProvider } from '@antv/graphin'
    <GraphinProvider>
         <NetworkTopologyViewer />
    </GraphinProvider>
    
// ---------OR initialize with data --------------
const graphinContextData = {...};
<GraphinProvider value={graphinContextData}>
   <NetworkTopologyViewer />
</GraphinProvider>
//-----------------------------------

/**
* @description
* Renders a Graph wrapped in a custom container
* and also sets up custom node behavior
*/
const NetworkTopologyViewer = () => {
  const { graph, api} = useGraphin(); ✅ does not error because this component is inside a Provider in a parent component
  const handleNodeClick = e => {.....})
  
  graph.on('node:click', handleNodeClick) // I can setup my node behavior outside of <Graphin> 🎉
  
	return (
	 <Card title="Network Topology">
		   <Graphin {...GraphinProps....}/>
      <Card>
  )
}
```

## Proposal
- make a new component called `GraphinProvider` that is exportable. 
   - By default if no data is passed into `GraphinProvider`  the data will be empty, but after `<Graphin>` component loads, it will update the `GraphinProvider`'s  context data
- make a hook `useGraphin` that allows the user to grab and modify data from the `GraphinProvider`
    - no more need to import `GraphinContext` since the `useGraphin` hook can get/set data from the `GraphinContext`
    - this pattern is very popular in newer react libraries like `react-query`, `react-router` etc...

## Example Fake Demo
Demo 1: https://stackblitz.com/edit/react-ts-cuhbx2?file=App.tsx
<img width="1231" alt="CleanShot 2021-12-05 at 17 50 22@2x" src="https://user-images.githubusercontent.com/6743796/144774762-32f52ed7-0960-41c3-a29b-3e69ce8a7cbd.png">
<img width="1628" alt="CleanShot 2021-12-05 at 17 55 56@2x" src="https://user-images.githubusercontent.com/6743796/144775155-48fce17c-7607-4b46-8ce8-90c79aee8f7c.png">

Demo2:
- this demonstrates the pattern demonstrated above with another very common API example
   - https://stackblitz.com/edit/react-ts-p8dg9n?file=App.tsx

**Comment by cliffordfajardo** [comment]
@pomelo-nwu,@zxc0328 


@Blakko & @timlrx - I've noticed you two have contributed seem to have used the `graphin` library, do you have any opinions on this 🙏?

**Comment by cliffordfajardo** [comment]
@timlrx @Anderson-Liu 
how have you setup custom events for node & edge clicks?

https://stackblitz.com/edit/react-ts-bgsybt?file=component%2Findex.ts

**Comment by pomelo-nwu** [comment]
@cliffordfajardo Thank you for your suggestions, we will think carefully and give some solutions. If you want, you can add my dingtalk and talk about it in detail 
![sg](https://user-images.githubusercontent.com/10703060/144963226-d69b5564-dff0-4cb2-a22d-6724f084507b.jpg)

**Comment by cliffordfajardo** [comment]
@pomelo-nwu - I've joined the group.


**Question(s)**
1. Is it ok if I post in English?
   - [Ding chat has auto translate so other people in there can translate the messages](https://www.linkedin.com/pulse/case-you-missed-10-must-know-features-dingtalk-lite-jessie-yu)
https://user-images.githubusercontent.com/6743796/145264358-45118ca7-6dae-416b-8f66-659ce25ecb80.mp4

**Comment by cliffordfajardo** [comment]
## Another Use Case Example

Below is another use case example of how having a `useGraphin` hook and a `<GraphinProvider>` is helpful.
Imagine I want to place a search component above my `<Graphin>` component like this:

<img width="1677" alt="CleanShot 2021-12-10 at 10 33 05@2x" src="https://user-images.githubusercontent.com/6743796/145624279-8fbdf9c0-4180-4569-8660-f1ed1bbda5f2.png">



```
        <Card title={headingTitle}>
          <CanvasSearchBar onInputChange={() => console.log('input changed')} /> 🔴 won't work because we don't have access to the `GraphinProvider` around this component

          <Graphin
            data={rawData}
            height={550}
            layout={{
              type: 'graph-with-hierachy',
            }}
            linkCenter
            defaultNode={{ type: 'rect' }}
            fitCenter
            fitView
            animate
          >
            <ActivateRelations />
          </Graphin>
        </Card>
```

**Comment by mxz96102** [comment]
Hi, I'm also interested in 'useGraphin' hook. Features such as isReady, APIs and etc. are so effective.

But in my opinion, it should be done with fewer side effects for graphin. So I came up with an idea that uses a global "Graphin pool" for Graphin hooks. 

For users who wants to useGraphin, they only need to append a global unique id to Graphin component, like <Graphin id="graph-1" /> and it can useGraphin("graph-1") anywhere. 

Instance of Graphin will register into Graphin pool when init and remove when destroy, with register event, context will directly avaliable for useGraphin. (It's like a pub/sub model), and for users, they don't need to know graphin context, only change they need to know is that they can assign id to graphin instance and use graphin api throgh id anywhere they need.

What's your opinion about this pattern, looking forwards to your reply~

**Comment by cliffordfajardo** [comment]
Hi @mxz96102 - I will share some thought's later today exploring concerning your idea.

> _Features such as isReady, APIs and etc. are so effective._
I agree

**Comment by timlrx** [comment]
It's a nice idea! Currently for motif, we are just [storing the ref from Graphin](https://github.com/cylynx/motif.gl/blob/7b7084929c4384ef0e562e3a5393fd6d53304d0b/packages/motif/src/containers/Explorer.tsx#L178) and passing it through to the other widget components using a [provider](https://github.com/cylynx/motif.gl/blob/7b7084929c4384ef0e562e3a5393fd6d53304d0b/packages/motif/src/containers/Explorer.tsx#L73). The other components can then `useContext` and have access to the full G6 api.  

A `useGraphin` would help simplify things.

**Comment by cliffordfajardo** [comment]
Hi @mxz96102 

> But in my opinion, it should be done with fewer side effects for graphin

I agree that the changes that are made to Graphin should not dramatically change the usage of the library.
@mxz96102 would you be able to share an example of what the code looks like in practice via a codesanbox, some visual or some pseudo code? I just want to make sure I understand your idea 😄 




For the longevity of the Graphin project,  I do believe its in the best interest of the library & it's community to migrate away from having the [`Graphin.tsx` file](https://github.com/antvis/Graphin/blob/master/packages/graphin/src/Graphin.tsx#L42) be a Class component and to rather have it be a Function component for a few reasons:
- Function components is the modern way to write React code & the React docs are moving to hooks first (see [new docs beta](https://beta.reactjs.org/reference#installation))
- By maintaining 2 types fo components in the codebase, we have to re-learn the nuance of class components (also how to write tests for them) and all their lifecycle methods every time we need to modify Graphin component, which isn't often but I know it will happen. 
- We can't take advantage of custom or reusable hook abstractions inside Graphin component since Class components cannot make use of hooks
- We likely won't be able to take advantage of upcoming react features
   - [React 18 for External Store Libraries](https://www.youtube.com/watch?v=oPfSC5bQPR8)
   - [React without Memoization](https://www.youtube.com/watch?v=lGEMwh32soc)
- the codebase will be harder to contribute to, isntead we should strive towards minimizing as many barriers of entry as possible so people who would like to contribute to Graphin aren't immediately turned off by dealing with Class components. People's personal time is valuable and this effects not only potential contributors but also existing contributors too

**Comment by mxz96102** [comment]
@cliffordfajardo I came up with an id sample here and just extend the component, hope can help you in someway.

https://codesandbox.io/s/hookedgraphin-6bbsr?file=/App.tsx

**Comment by cliffordfajardo** [comment]
@mxz96102 - can you share more about your thoughts on the pros and cons of this approach versus other alternatives we've talked about in this discussion?

Genuinely, I am just curious and am hoping to understand your idea better so we can all create awesome things

**Comment by mxz96102** [comment]
I agree that deprecate class component will embrace more features of React 18, but for now, it might not be done in this version, considering that some of the users still depend on class lifecycle to do something, in this version using id might be enough for users, and for break change (like converting Graphin to function component), may happen in a new version like Graphin 3?
